### ISSUE TITLE
ArgumentError is reported at wrong level for some jitted methods

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6335]
When compiling a Ruby method to a fixed-arity form, we do not properly include that method in an ArgumentError resulting from arity mismatch.

For example:

```
$ jruby -e 'def foo(a); end; foo(1, 2)'
ArgumentError: wrong number of arguments (given 2, expected 1)
  <main> at -e:1

$ rvm ruby-2.5.7 do ruby -e 'def foo(a); end; foo(1, 2)'
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `foo': wrong number of arguments (given 2, expected 1) (ArgumentError)
```

Note that JRuby reports the error from `-e:1`, not from within the `foo` method. This is because when we can see that there's a single arity possible for a method, we will compile the JVM bytecode to receive exactly that number of arguments, and the arity-checking happens *outside* the method body as part of the dispatch logic.

You can see from the partial trace below that the error is raised from a synthetic "VARARGS" version of the method, whose sole purpose is to be the fallback path when the incoming arity does not match the method we generated in bytecode:

```
ArgumentError: wrong number of arguments (given 2, expected 1)
...
             raiseArgumentError at org/jruby/runtime/Arity.java:255
                     checkArity at org/jruby/ir/runtime/IRRuntimeHelpers.java:595
                     checkArity at org/jruby/ir/targets/indy/Bootstrap.java:290
  RUBY$method$foo$0$__VARARGS__ at -e:-1
                           call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:109
                           call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:208
                   cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:396
                           call at org/jruby/runtime/callsite/CachingCallSite.java:205
               invokeOther0:foo at -e:1
                         <main> at -e:1

```

And that frame gets explicitly ignored, since it's just for redispatching or raising an error, at this point:

https://github.com/jruby/jruby/blob/2406b50fdacc29e332d00d06cb24481a76f1db38/core/src/main/java/org/jruby/util/JavaNameMangler.java#L318

When running this code with the interpreter, the arity check happens *inside* the method body, so the trace matches:

```
$ jruby -X-C -e 'def foo(a); end; foo(1, 2)'
ArgumentError: wrong number of arguments (given 2, expected 1)
     foo at -e:1
  <main> at -e:1
```

There's a couple possible fixes here:

1. Allow the `VARARGS` version of the method to show up in the trace. Possibly only when raising an ArgumentError from that point, which would be difficult to determine.
2. Raise the error slightly differently from the `VARARGS` version, so that the original file and line can be forcibly added to the trace. This will be cumbersome and probably error-prone to maintain.
3. Emit an additional method body that is not filtered by exception tracing and which is *only* used by the `VARARGS` version to raise the error. This would replace the specific-arity frame in the trace and we would no nothing ever calls through it, so we would not need the filtering in option 1.

I think option 3 would be the least invasive, but it means emitting an additional method body for every place we create a `VARARGS` wrapper now.

**Comment by headius** [comment]
Note this error has been around for a very, very long time, and also affects most core class methods (since we usually will check call arity *before* dispatching to them:

```
$ jruby -e '"foo".split(1, 2, 3)'
ArgumentError: wrong number of arguments calling `split` (given 3, expected 0..2)
  <main> at -e:1
```

It's not a major issue, in my estimation, since we only lose the frame that's associated with the target method, and you can still see the frame indicating where the bad call was made. However locating the method body that *would have* been called can be confusing if that frame gets dropped.

**Comment by headius** [comment]
Fixed by #6771.
